### PR TITLE
Fixes serialization roundtrip of enums in eve (issue #352)

### PR DESCRIPTION
**Pull request by tehrengruber**
## Description

@egparedes The issue lied in ours validators not Pydantic...

## Requirements

Before submitting this PR, please make sure:

- [x] You have run the code checks, tests and documentation build successfully
- [x] All fixes and all new functionality are tested and documentation is up to date
- [x] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/master/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Comment by tehrengruber** [comment]
bors try

File: src/eve/type_definitions.py [hunk 1]
```diff
@@ -56,41 +56,43 @@ class Enum(enum.Enum):
...
+        yield cls._value_validator
 
     @classmethod
-    def _strict_type_validator(cls, v: Any) -> StrEnum:
-        if not isinstance(v, cls):
-            raise TypeError(f"Invalid value type [expected: {cls}, received: {v.__class__}]")
-        return v
+    def _value_validator(cls, v: Any) -> StrEnum:
+        if hasattr(cls, v):
+            v = cls[v]
```
Comment (egparedes) [inline src/eve/type_definitions.py:92]: Either `v` is guaranteed to be a `str` and thus the type annotation is wrong or the call to `hasattr` can trigger an Exception if something else than a string is passed. I think the safe way to check if the passed value is a string with the name of an enum member is using the `__members__` dictionary of the enum class:

```suggestion
    def _value_validator(cls, v: Any) -> StrEnum:
        if isinstance(v, str) and v in cls.__members__:
            v = cls[v]
```

It's probably worth to add a comment here mentioning that the passed string is first looked up as a enum member and then as a enum value.

File: tests/eve_tests/definitions.py [hunk 2]
```diff
@@ -361,7 +361,7 @@ def make_simple_node_with_abstract_collections(
     fixed: bool = False,
 ) -> SimpleNodeWithAbstractCollections:
     int_value = make_int_value(fixed=fixed)
-    int_sequence = make_collection_value(int, collection_type=tuple, length=3)
```
Comment (egparedes) [inline tests/eve_tests/definitions.py:364]: I don't understand this change. Why is it needed?

**Review by egparedes — CHANGES_REQUESTED** [review]
Great simple fix for the long-standing issue, thanks for looking into that. I only have a question and a quick fix.

**Comment by tehrengruber** [comment]
bors try

File: src/eve/type_definitions.py [hunk 3]
```diff
@@ -56,41 +56,43 @@ class Enum(enum.Enum):
...
-        yield cls._strict_type_validator
+        yield cls._value_validator
 
     @classmethod
-    def _strict_type_validator(cls, v: Any) -> StrEnum:
-        if not isinstance(v, cls):
-            raise TypeError(f"Invalid value type [expected: {cls}, received: {v.__class__}]")
-        return v
+    def _value_validator(cls, v: Any) -> StrEnum:
+        if hasattr(cls, v):
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/eve/type_definitions.py]: What does pydantic uses as default for the serialization: the name or the value? If it always uses of them, we should also probably use the same

**Comment by tehrengruber** [comment]
bors try

File: src/eve/type_definitions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/eve/type_definitions.py]: Pydantic always uses the value for serialization (for `dict` it is configurable whether the enum value or instance is used). I removed the `str` case.

**Review by egparedes — APPROVED** [review]
Looks good.

**Pull request merged**